### PR TITLE
fixes #26955 - load boot settings from env variables

### PR DESCRIPTION
**Pull request by timogoebel**
This commit adds support for defining the `SETTINGS` via env variables.

```sh
FOREMAN_HSTS_ENABLED=false bundle exec rails server
FOREMAN_LOGGERS_FOREMAN_WRECKINGBALL__IMPORT_ENABLED=false bundle exec rails server
```

### CODE REVIEW

**Review by ohadlevy — DISMISSED** [review]
I like this a lot more than continuing on extending the settings file, the main two points are:

- how do we keep it in sync? does it make sense to add something like permissions tests that if you add a setting a test will fail?
- where do we document it?

**Comment by timogoebel** [comment]
> * how do we keep it in sync? does it make sense to add something like permissions tests that if you add a setting a test will fail?

Do we have a list of all settings that are supported in the current settings file? I think we don't. So I don't know with what we can compare the values from the environment.

**Comment by ohadlevy** [comment]
> Do we have a list of all settings that are supported in the current settings file? I think we don't. So I don't know with what we can compare the values from the environment.

Well, until now we had only one place where we had settings define, now two, but I'm OK to merge as is and move the responsibility for maintainers to keep in sync?

**Comment by timogoebel** [comment]
> Well, until now we had only one place where we had settings define, now two, but I'm OK to merge as is and move the responsibility for maintainers to keep in sync?

We could do it the over way around. If the settings hash contains data that is not available via environment variables, we could print a warning.
We also need to find a solution for settings files for plugins.

**Comment by ohadlevy** [comment]
> We also need to find a solution for settings files for plugins.

Can't we reuse  the same code that goes across all <FOREMAN... > env variables?

**Comment by timogoebel** [comment]
> Can't we reuse the same code that goes across all <FOREMAN... > env variables?

This just works for the log settings as we can't cast the values automatically. I'd be ok having this for core settings to begin with, we can add a plugin api later. Afaik only foreman_memcache and [katello](https://github.com/Katello/katello/blob/master/config/katello.yaml.example) uses this.

**Comment by ohadlevy** [comment]
> > Can't we reuse the same code that goes across all <FOREMAN... > env variables?
> 
> This just works for the log settings as we can't cast the values automatically. I'd be ok having this for core settings to begin with, we can add a plugin api later. Afaik only foreman_memcache and [katello](https://github.com/Katello/katello/blob/master/config/katello.yaml.example) uses this.

I agree, for memcache, redis should invalidate it, so ack.
from quick look on katello's I can't see a real reason why its required for application boot? in general settings (not the Settings Model) should only affect application start up until the setting model is available. @ehelms @jlsherrill ?

**Comment by ehelms** [comment]
I'm don't follow the question. Are you asking why Katello uses SETTINGS vs. settings?

**Comment by ehelms** [comment]
My first thought was do we need to define every setting to manage it via an environment variable? Could we do it lazily (until there is an API for defining settings perhaps) and check for any `FOREMAN_` within the `env` and merge those into SETTINGS at the end letting them take precedence?

For Katello, some of whats in the settings file is historically there to prevent 'simple' changes by an admin givne they have profound impacts. We've not re-evaluated those in a long time and some could be moved to hard-coded, others to Settings model and some to new architecture choices with the smart proxy.

**Comment by ohadlevy** [comment]
> My first thought was do we need to define every setting to manage it via an environment variable? Could we do it lazily (until there is an API for defining settings perhaps) and check for any `FOREMAN_` within the `env` and merge those into SETTINGS at the end letting them take precedence?
> 
> For Katello, some of whats in the settings file is historically there to prevent 'simple' changes by an admin givne they have profound impacts. We've not re-evaluated those in a long time and some could be moved to hard-coded, others to Settings model and some to new architecture choices with the smart proxy.

thanks for the follow up, based on that, if we do introduce this patch, any concerns for katello (or rather katello will simply add similar code for those settings, or introduce refactoring if needed?

thanks!

**Comment by timogoebel** [comment]
> My first thought was do we need to define every setting to manage it via an environment variable? Could we do it lazily (until there is an API for defining settings perhaps) and check for any `FOREMAN_` within the `env` and merge those into SETTINGS at the end letting them take precedence?

That would be great, however, I do prefer an explicit approach here. We can definitely figure out the path where a setting would need to be stored inside the `SETTINGS` hash by parsing the environment variable names, but we can't do any casting. How would we know that `http://example.com http://www.example.com` is meant to be an array and not a string?

In general, I believe this patch won't hurt. We could just not override settings introduced in plugins.

File: config/settings.rb
Comment (ohadlevy) [inline config/settings.rb]: is this suppose to be Boolean?

**Comment by ohadlevy** [comment]
tested the patch with the following docker-compose file changes successfully (meaning you can apply it and get the same environment)
```diff
diff --git a/docker-compose.yml b/docker-compose.yml
index 821984ff2..576b52800 100644
--- a/docker-compose.yml
+++ b/docker-compose.yml
@@ -20,7 +20,7 @@ services:
       - db:/var/lib/mysql/data
 
   app: &app_base
-    image: quay.io/foreman/foreman:develop
+    image: quay.io/ohadlevy/foreman:6826
     command: bundle exec bin/rails server -b 0.0.0.0
     build:
       context: .
@@ -28,6 +28,9 @@ services:
       - DATABASE_URL=mysql2://foreman:foreman@db:3306/foreman
       - RAILS_MAX_THREADS=5
       - RAILS_ENV=production
+      - FOREMAN_MARK_TRANSLATED=true
+      - FOREMAN_FQDN=foreman.example.com
+      - FOREMAN_DOMAIN=example.com
     hostname: foreman.example.com
     links:
       - db
```

**Comment by ohadlevy** [comment]
@timogoebel could you please add setting to expose redis as well? thanks.

**Comment by timogoebel** [comment]
@ohadlevy: Done.

**Review by ekohl — COMMENTED** [review]
Would it be possible to have some tests for this? Especially when there's multiple arguments. Perhaps some documentation?

**Comment by timogoebel** [comment]
> Would it be possible to have some tests for this?

Sure, somebody would just need to write them.

> Perhaps some documentation?

What kind of documentation do you have in mind?

**Comment by timogoebel** [comment]
> > Would it be possible to have some tests for this?
> 
> Sure, somebody would just need to write them.
> 

I started to work on some tests, please wait before merging.

File: config/settings.rb
Comment (ekohl) [inline config/settings.rb]: I'm having my doubts about this. Currently this must be set in `boot_settings` since that one is included in `Gemfile`. Setting it this late just breaks it if they don't match.

This at least needs to set the right variable :)

```suggestion
  'FOREMAN_RAILS' => [:boolean, :rails],
```

File: config/settings.rb
Comment (ekohl) [inline config/settings.rb]: Does this one really make sense? Can't we just check if jsonp can be loaded and automatically determine this?

I'd think that the following would raise an exception if it can't be loaded:

https://github.com/theforeman/foreman/blob/c5fbd28993fbe29f3457a329245e8975dd4c7667/config/application.rb#L68-L74

Otherwise we can check if `Rack::JSONP` is defined here:

https://github.com/theforeman/foreman/blob/c5fbd28993fbe29f3457a329245e8975dd4c7667/config/application.rb#L154-L155

Doing this would actually allow us to get rid of the setting altogether.

**Review by ekohl — CHANGES_REQUESTED** [review]
> What kind of documentation do you have in mind?

Right now nobody knows how to find out this is even possible. Perhaps a note on where to find the list of possible env vars would be sufficient? It'd also be good to know who the audience is to know where to document this.

File: config/settings.rb
Comment (timogoebel) [reply config/settings.rb]: Yeah, found that as well while writing the tests. ;-)

File: config/settings.rb
Comment (timogoebel) [reply config/settings.rb]: I think we can get rid of this setting, but that's unrelated to this patch. If we want to remove it, let's do it in a separate PR.

**Comment by timogoebel** [comment]
@ekohl: Extracted this to a small helper class, fixed the rails boot setting, added a couple of tests.

I believe we should document this in the normal way, e.g. the manual.

File: app/services/foreman/env_settings_loader.rb
Comment (ohadlevy) [inline app/services/foreman/env_settings_loader.rb:31]: do you expect plugins to extend this list?

File: config/settings.rb
Comment (ekohl) [reply config/settings.rb]: While I agree it should be a separate effort, I also wonder if including it in this list implies support. I'd consider not exposing the env var. The installer doesn't expose the setting and it defaults to false. That makes me wonder who uses this. Feels to me that jsonp is a technique that's no longer relevant and we could deprecate it.

Comment (ohadlevy) [reply config/settings.rb]: AFAIR this was introduced by ovirt project, maybe @masayag knows if its still being used. regardless its outside the scope of this PR (to remove it) and its pretty safe to not include it (As an env variable) if we don't want to encourage people using it.

**Review by ohadlevy — APPROVED** [review]

**Review by ekohl — APPROVED** [review]
I'd still consider leaving out `FOREMAN_RAILS` and `FOREMAN_SUPPORT_JSONP` since we don't want to encourage users to set these but I won't block on that

File: config/settings.rb
Comment (masayag) [reply config/settings.rb]: @ohadlevy I'm not familiar of any dependency of ovirt_provision_plugin on that set of env-variables.

Comment (ohadlevy) [reply config/settings.rb]: @masayag sorry I wasn't clear enough, the question here is if ovirt uses the JSONP feature of foreman.

Comment (masayag) [reply config/settings.rb]: @ohadlevy yes, the entire interaction (requests & responses) from ovirt's side is done via JSON, i.e. `/api/v2/hosts?format=json&per_page=9999` is notation that states the JSONP feature is used.

https://github.com/oVirt/ovirt-engine/blob/master/backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/host/provider/foreman/ForemanHostProviderProxy.java#L37

Comment (ohadlevy) [reply config/settings.rb]: @masayag json and jsonp are not the same... see https://www.w3schools.com/js/js_json_jsonp.asp

Comment (ohadlevy) [reply config/settings.rb]: perhaps @ofavre remembers? :)

**Comment by ohadlevy** [comment]
thanks @timogoebel

**Pull request merged**
